okta/resource_users.go

```diff
@@ -110,126 +240,213 @@ func resourceUserRead(d *schema.ResourceData, m interface{}) error {
-		userRoles, _, err := client.Users.ListRoles(userList.ID)
-		if err != nil {
-			return fmt.Errorf("[ERROR] Error listing user role: %v", err)
-		}
-		switch {
-
-		case userRoles == nil && d.Get("role").(string) != "":
-			log.Printf("[INFO] Assigning role: " + d.Get("role").(string))
-			_, err = client.Users.AssignRole(userList.ID, d.Get("role").(string))
-			if err != nil {
-				return fmt.Errorf("[ERROR] Error assigning role to user: %v", err)
-			}
-
-		case userRoles != nil && d.Get("role").(string) != "":
-			log.Printf("[INFO] Changing role: " + d.Get("role").(string) + " to " + userRoles.Role[0].Type)
-			_, err = client.Users.UnAssignRole(userList.ID, userRoles.Role[0].ID)
-			if err != nil {
-				return fmt.Errorf("[ERROR] Error removing role from user: %v", err)
-			}
-			_, err = client.Users.AssignRole(userList.ID, d.Get("role").(string))
-			if err != nil {
-				return fmt.Errorf("[ERROR] Error assigning role to user: %v", err)
-			}
-
-		case userRoles != nil && d.Get("role").(string) == "":
-			log.Printf("[INFO] Removing role: " + d.Get("role").(string))
-			_, err = client.Users.UnAssignRole(userList.ID, userRoles.Role[0].ID)
-			if err != nil {
-				return fmt.Errorf("[ERROR] Error removing role from user: %v", err)
-			}
-
-		default:
-			return fmt.Errorf("User role changed but Terraform was unable to apply. Please investigate.")
-		}
+	userTemplate("update", d, m)
+	if err != nil {
+		return err
 	}
+	d.Partial(false)
 
 	return nil
 }
 
 func resourceUserDelete(d *schema.ResourceData, m interface{}) error {
-	log.Printf("[INFO] Delete User %v", d.Get("email").(string))
+	log.Printf("[INFO] Delete User %v", d.Get("login").(string))
 	client := m.(*Config).oktaClient
 
-	userList, _, err := client.Users.GetByID(d.Get("email").(string))
+	userList, _, err := client.Users.GetByID(d.Id())
 	if err != nil {
 		return fmt.Errorf("[ERROR] Error GetByID: %v", err)
 	}
-
 	// must deactivate the user before deletion
 	if userList.Status != "DEPROVISIONED" {
-		_, err := client.Users.Deactivate(d.Get("email").(string))
+		_, err := client.Users.Deactivate(d.Id())
 		if err != nil {
 			return fmt.Errorf("[ERROR] Error Deactivating user: %v", err)
 		}
 	}
-
-	// now! delete the user
-	deleteUser, err := client.Users.Delete(d.Get("email").(string))
+	// delete the user
+	_, err = client.Users.Delete(d.Id())
 	if err != nil {
 		return fmt.Errorf("[ERROR] Error Deleting user: %v", err)
 	}
-	log.Printf("[INFO] Okta User Deleted: %+v", deleteUser)
-
 	// delete the user resource from terraform
 	d.SetId("")
 
 	return nil
 }
+
+func userTemplate(action string, d *schema.ResourceData, m interface{}) error {
+	client := m.(*Config).oktaClient
+
+	template := client.Users.NewUser()
+	template.Profile.Login = d.Get("login").(string)
+	template.Profile.Email = d.Get("login").(string)
+	template.Profile.FirstName = d.Get("firstname").(string)
+	template.Profile.LastName = d.Get("lastname").(string)
+	if _, ok := d.GetOk("email"); ok {
```

Comment:
Would it be possible to build a static array of fields and then iterate over that to make this more compact?
Something like (note this is pseduo-code) :
```
fields = ["middlename", "secondemail", "honprefix"]

for field in fields {
    if _, ok := d.GetOk(field); ok {
        template.Profile.get_attribute_reference(field) = d.Get(field).(string)
    }
}
```